### PR TITLE
[TAP] Offline Notes PR

### PR DESCRIPTION
**Pull request by kwright29**
# Description
### Goal
- Allows notes to be downloaded into local datastore
### Offline Notes:
- Created data model using CoreData
- Created entity within data model mirroring Note object 
- Added download button to Feed view
- Gives alert message if note is successfully saved or if there is an error saving 

# Tests

### Tests Included:
- Downloading a note w/o PFFileObject data 
   - throws an error if PFFileObject is nil
- Downloading a note w/ PFFileObject data

### [Test Video](https://pxl.cl/295c5)

# What's Next?
- Creating profile view
- Creating fetch request to display downloaded notes on profile view
- Deleting, updating and syncing notes in CoreData

### CODE REVIEW

File: NoteFind/Alerts/ErrorAlerts.m [hunk 1]
```diff
@@ -108,4 +108,26 @@ + (void)errorPostingNote:(NewNoteViewController *)newNoteVC {
...
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) { /*handle response here */ }];
+    [alert addAction:okAction];
+    
+    [vc presentViewController:alert animated:YES completion:^{
+        
+    }];
+}
+
++ (void)successDownloading:(UIViewController *)vc {
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"success!" message:@"this note was successfully downloaded! check your profile for all of your downloaded notes." preferredStyle:(UIAlertControllerStyleAlert)];
```
Comment (rahul-malik) [inline NoteFind/Alerts/ErrorAlerts.m:124]: This can be in another PR but do you think it might make sense to create a helper function for all these dialogs since the code looks pretty identical for each one?

File: NoteFind/Delegates/AppDelegate.h [hunk 2]
```diff
@@ -6,11 +6,12 @@
...
 
 #import <UIKit/UIKit.h>
-
+#import <CoreData/CoreData.h>
 
 @interface AppDelegate : UIResponder <UIApplicationDelegate>
+@property (nonatomic, strong) NSPersistentContainer *persistentContainer;
 
-//@property (nonatomic, strong, nullable) id<OIDExternalUserAgentSession> currentAuthorizationFlow;
+- (void)saveContext;
```
Comment (rahul-malik) [inline NoteFind/Delegates/AppDelegate.h:14]: Leave a comment here explaining what `saveContext` does

File: NoteFind/Delegates/AppDelegate.h [hunk 3]
```diff
@@ -6,11 +6,12 @@
 //
 
 #import <UIKit/UIKit.h>
-
+#import <CoreData/CoreData.h>
 
 @interface AppDelegate : UIResponder <UIApplicationDelegate>
+@property (nonatomic, strong) NSPersistentContainer *persistentContainer;
```
Comment (rahul-malik) [inline NoteFind/Delegates/AppDelegate.h:12]: you probably want this to also be `readonly` since you wouldn't want another part of the app to change the NSPersistentContainer created in the App Delegate

File: NoteFind/Delegates/AppDelegate.h [hunk 4]
```diff
@@ -6,11 +6,12 @@
 //
 
 #import <UIKit/UIKit.h>
-
+#import <CoreData/CoreData.h>
```
Comment (rahul-malik) [inline NoteFind/Delegates/AppDelegate.h:9]: You can move this import to AppDelegate.m and just forward declare the NSPersistentContainer here

```
@class NSPersistentContainer;
```

File: NoteFind/Delegates/AppDelegate.m [hunk 5]
```diff
@@ -6,15 +6,17 @@
 //
 
 #import "AppDelegate.h"
-#import "Parse/Parse.h"
+#import "ErrorAlerts.h"
+#import "FeedViewController.h"
+#import <Parse/Parse.h>
```
Comment (rahul-malik) [inline NoteFind/Delegates/AppDelegate.m:11]: split out third party framework imports into a separate sorted group

File: NoteFind/Models/Note.m [hunk 6]
```diff
@@ -38,7 +38,7 @@ + (void) postUserNote: (UIImage * _Nullable)image withDescription: (NSString * _
     newNote.noteDescription = description;
     newNote.addCount = @(0);
     newNote.title = title;
-
+
```
Comment (rahul-malik) [inline NoteFind/Models/Note.m:41]: You can undo this so the file isn't included in your PR

File: NoteFind/View Controllers/FeedViewController.h [hunk 7]
```diff
@@ -12,11 +12,14 @@
 #import "LoginViewController.h"
 #import "Note.h"
 #import "NoteCell.h"
+#import <CoreData/CoreData.h>
```
Comment (rahul-malik) [inline NoteFind/View Controllers/FeedViewController.h:15]: Forward declare `NSPersistentContainer`

File: NoteFind/Views/NoteCell.h [hunk 8]
```diff
@@ -12,13 +12,17 @@
 #import "Parse/PFImageView.h"
 
 NS_ASSUME_NONNULL_BEGIN
+@protocol DataFailureDelegate
+- (void)showErrorMessage;
```
Comment (rahul-malik) [inline NoteFind/Views/NoteCell.h:16]: Nice job setting up the communication between the table view cell and table view itself!

Comment (rahul-malik) [inline NoteFind/Views/NoteCell.h:16]: `showErrorMessage` seems pretty generic, could we use a more specific name here?

File: NoteFind/Views/NoteCell.m [hunk 9]
```diff
@@ -5,13 +5,24 @@
 //  Created by Khloe Wright on 7/8/22.
 //
 
+#import "AppDelegate.h"
+#import <CoreData/CoreData.h>
 #import "NoteCell.h"
 
+
+@interface NoteCell () {
+    AppDelegate *appDelegate;
```
Comment (rahul-malik) [inline NoteFind/Views/NoteCell.m:14]: We shouldn't keep a reference to the AppDelegate / Context here since you can always access it from `[UIApplication sharedApplication].delegate`.

Additionally this sets up a retain cycle / memory leak since 
AppDelegate retains Feed which retains Note Cells which now retains the AppDelegate

Comment (rahul-malik) [reply NoteFind/Views/NoteCell.m:14]: An alternative would also to just make both of these instance variables `weak` in order to make sure we aren't holding a strong reference to them

File: NoteFind/Views/NoteCell.m [hunk 10]
```diff
@@ -24,9 +35,33 @@ - (void)setNote:(Note *)note {
...
+        [appDelegate saveContext];
+        
+    }
+    // TODO: add addCount
+    
+    // TODO: fetch data
+    
+    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"OfflineNote"];
+    NSArray *results = [context executeFetchRequest:request error:nil];
+    NSLog(@"results: %@", results);
```
Comment (rahul-malik) [inline NoteFind/Views/NoteCell.m:64]: What are these past few lines used for? Seems like debugging - if so lets remove it from your commiit

File: NoteFind/Views/NoteCell.m [hunk 11]
```diff
@@ -24,9 +35,33 @@ - (void)setNote:(Note *)note {
...
+    _postUsername.text = _note.author.username;
     _postNote.file = _note.note;
     [_postNote loadInBackground];
+    
+    // TODO: add addCount
+}
+- (IBAction)didTapDownload:(id)sender {
+    NSManagedObject *newOfflineNote = [NSEntityDescription insertNewObjectForEntityForName:@"OfflineNote" inManagedObjectContext:context];
+    NSData *fileData = [_postNote.file getData];
+    if (fileData == nil) {
```
Comment (rahul-malik) [inline NoteFind/Views/NoteCell.m:47]: How do we handle notes that we've already downloaded to make sure there aren't duplicate objects in CoreData?

File: Podfile [hunk 12]
```diff
@@ -8,10 +8,9 @@ target 'NoteFind' do
   # Pods for NoteFind
 pod 'Parse'
 pod 'Parse/UI'
-pod 'DateTools'
-pod 'NotionClient'
-pod 'AppAuth'
 pod 'AFNetworking'
+pod 'Parse+NSCoding', '~> 0.1.4'
```
Comment (rahul-malik) [inline Podfile:12]: Did you end up using this library? If not I'd remove it and rebuild your pods to avoid checking in more source files than you have to

**Review by rahul-malik — CHANGES_REQUESTED** [review]

File: NoteFind/Delegates/AppDelegate.h (hunk 2 above)
Comment (kwright29) [reply NoteFind/Delegates/AppDelegate.h:14]: done!

File: NoteFind/Delegates/AppDelegate.h (hunk 3 above)
Comment (kwright29) [reply NoteFind/Delegates/AppDelegate.h:12]: done!

File: NoteFind/Delegates/AppDelegate.h (hunk 4 above)
Comment (kwright29) [reply NoteFind/Delegates/AppDelegate.h:9]: done!

File: NoteFind/Delegates/AppDelegate.m (hunk 5 above)
Comment (kwright29) [reply NoteFind/Delegates/AppDelegate.m:11]: done!

File: NoteFind/View Controllers/FeedViewController.h (hunk 7 above)
Comment (kwright29) [reply NoteFind/View Controllers/FeedViewController.h:15]: i meant to take this out i don't use the container so i'll delete that line & the container

File: NoteFind/Views/NoteCell.h (hunk 8 above)
Comment (kwright29) [reply NoteFind/Views/NoteCell.h:16]: thank you!

Comment (kwright29) [reply NoteFind/Views/NoteCell.h:16]: done!

File: NoteFind/Views/NoteCell.m (hunk 9 above)
Comment (kwright29) [reply NoteFind/Views/NoteCell.m:14]: okay, since i do everything pertaining offline notes in the downloading, i'll put both variables in there

Comment (kwright29) [reply NoteFind/Views/NoteCell.m:14]: ^ didTapDownload

File: NoteFind/Views/NoteCell.m (hunk 11 above)
Comment (kwright29) [reply NoteFind/Views/NoteCell.m:47]: fixed- i made an offline data manager to handle all of that and moved my core data code to that file.

File: Podfile (hunk 12 above)
Comment (kwright29) [reply Podfile:12]: done!

File: NoteFind/Alerts/ErrorAlerts.m (hunk 1 above)
Comment (kwright29) [reply NoteFind/Alerts/ErrorAlerts.m:124]: definitely! i'll include this on another PR

File: NoteFind/View Controllers/FeedViewController.h [hunk 13]
```diff
@@ -13,10 +13,13 @@
 #import "Note.h"
 #import "NoteCell.h"
 
+
 NS_ASSUME_NONNULL_BEGIN
 
 @interface FeedViewController : UIViewController
 
+
```
Comment (rahul-malik) [inline NoteFind/View Controllers/FeedViewController.h:21]: maybe just remove whitespace changes?

File: NoteFind/Views/NoteCell.m [hunk 14]
```diff
@@ -5,13 +5,19 @@
 //  Created by Khloe Wright on 7/8/22.
 //
 
+#import "AppDelegate.h"
+#import "OfflineDataManager.h"
 #import "NoteCell.h"
 
+
+@interface NoteCell ()
```
Comment (rahul-malik) [inline NoteFind/Views/NoteCell.m:13]: Do we need this category since it doesn't do anything?

**Review by rahul-malik — APPROVED** [review]

File: NoteFind/Delegates/AppDelegate.m [hunk 15]
```diff
@@ -49,5 +50,64 @@ - (void)application:(UIApplication *)application didDiscardSceneSessions:(NSSet<
...
+@synthesize persistentContainer = _persistentContainer;
+
+- (NSPersistentContainer *)persistentContainer {
+    // The persistent container for the application. This implementation creates and returns a container, having loaded the store for the application to it.
+    @synchronized (self) {
+        if (_persistentContainer == nil) {
+            _persistentContainer = [[NSPersistentContainer alloc] initWithName:@"OfflineData"];
+            [_persistentContainer loadPersistentStoresWithCompletionHandler:^(NSPersistentStoreDescription *storeDescription, NSError *error) {
+                if (error != nil) {
+                    // Replace this implementation with code to handle the error appropriately.
```
[the remark this answers is not in the document]
Comment (kwright29) [reply NoteFind/Delegates/AppDelegate.m:64]: done

File: NoteFind/OfflineData/OfflineDataManager.m [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+//  OfflineDataManager.m
+//  NoteFind
+//
+//  Created by Khloe Wright on 8/3/22.
+//
+#import "AppDelegate.h"
+#import "ErrorAlerts.h"
+#import "OfflineDataManager.h"
+#import "PFFileObject.h"
+#import <Parse/Parse.h>
```
[the remark this answers is not in the document]
Comment (kwright29) [reply NoteFind/OfflineData/OfflineDataManager.m:11]: done!

File: NoteFind/Views/NoteCell.m (hunk 14 above)
Comment (kwright29) [reply NoteFind/Views/NoteCell.m:13]: done!
